{% load static %}

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">

<head>
    <title>Vaxign2: Vaccine Design</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta http-equiv="Content-Script-Type" content="text/javascript" />
    <link rel="shortcut icon" href="/favicon.ico"/>
    <link href="{% static 'css/bmain.css' %}" rel="stylesheet" type="text/css" />
    <link href="{% static 'css/vaxign_home.css' %}" rel="stylesheet" type="text/css" />
</head>

<body style="margin:0px;" id="main_body">

{% include 'template_vaxign_top.html' %}  

<table width="100%" border="0" cellpadding="0" cellspacing="0">
    <tr>
        <td width="160" valign="top">
            {% include 'template_vaxign_left.html' %}
        </td>
        <td valign="top">
            {% if 'c_user_name' in request.session %}
                <div style="text-align:right; margin: 2px 8px 2px 8px;">            
                    {% include 'template_small_navi.html' %}
                </div>
            {% endif %}
            <div class="home_container">
                <h3 align="center">The Vaxign2 Tutorial</h3>

                <p>&nbsp;</p>
                <p><strong><u>Notes: </u></strong></p>
                <ul>
                <li><strong>The following tutorial comes from the Vaxign Demo material prepared for the <a href="http://www.violinet.org/icovax2012">ICoVax 2012 Workshop. </a></strong></li>
                <li><strong>This material was prepared by Allen Xiang and Oliver He for Vaxign2 on October 6, 2012, and last updated by Michael F Cooke on February 15, 2021 for Vaxign2.</strong></li>
                </ul>
                <p>&nbsp;</p>
                <p><strong>Introduction:</strong> Vaxign2 is a web-based software program in VIOLIN that targets for vaccine design. Based on the reverse vaccinology strategy, Vaxign2 predicts vaccine targets by bioinformatics analysis of genome sequences. Genome sequences come from pathogenic strains, non-pathogenic strains (optional), and host species (human, mouse, or pig). Predicted features in the Vaxign2 pipeline include protein subcellular location, transmembrane helices, adhesin probability, conservation among pathogenic strains, sequence exclusion  from genomes of non-pathogenic strains, sequence similarity to host proteins, and epitope binding to MHC class I and class II. Vaxign2 contains precomputed predictions for over 398 genomes and also allows dynamic vaccine target prediction based on users’ input sequences.</p>
                <p>
                The following tutorial will provide step-by-step instructions on how to navigate through the Vaxign2 website (<a href="http://www.violinet.org/vaxign2">http://www.violinet.org/vaxign2</a>). Two use cases will be tested: (1) Human SARS coronavirus 2 (SARS-CoV-2) isolate Wuhan-Hu-1 vaccine target prediction; (2) Enterohemorrhagic <em>Escherichia coli</em> (EHEC) O157:H7 vaccine target prediction. The SARS-CoV-2 use case will primarily demonstrate Vaxign2 sequence conservation analyses and MHC class I epitope prediction techniques. The O157:H7 use case will primarily demonstrate general features for bacterial vaccine target prediction including prediction of secreted and outer membrane proteins and transmembrane helices. Both tutorials will use pre-computed results to demonstrate the feature and save time. This tutorial will also demonstrate how to set up your own VIOLIN account to save query results and facilitate collaborative analysis with others.</p>
                <p>&nbsp;</p>

                <p><strong><u>First use case: SARS-CoV-2 Vaccine Target Prediction</u></strong></p>
                <p><strong>Visit the Vaxign2 job submission page.</strong> Visit <a href="http://www.violinet.org/vaxign2">http://www.violinet.org/vaxign2</a> in a modern web browser of your choice. Vaxign2 is freely available for public use without any requirement to sign a license or create an account; Creating a VIOLIN account to save your query results for later personal or collaborative use is completely optional. The Vaxign2 job submission page (Figure 1) allows you to submit Vaxign2 job by uploading or entering your own sequence input for <strong>Dynamic Analysis</strong>, or by choosing a genome that has been pre-processed using the <strong>Precompute Query</strong> tab (Figure 2).</p>

                <div align="center">
                    <figure>
                        <a href="{% static "tutorial/figure1.png" %}"><img src="{% static "tutorial/figure1.png" %}" alt="Figure 1: Vaxign2 job submission page" width="800" /></a>
                        <figcaption>Figure 1: Vaxign2 job submission page</figcaption>
                    </figure>
                    <figure>
                        <a href="{% static "tutorial/figure2.png" %}"><img src="{% static "tutorial/figure2.png" %}" alt="Figure 2: Vaxign2 precomputed queries tab" width="800" /></a>
                        <figcaption>Figure 2: Vaxign2 precomputed queries tab</figcaption>
                    </figure>
                </div>
                <p>&nbsp;</p>

                <p><strong>Select a Genome Group.</strong> In the Precompute Query tab, click the “Select a Genome Group” dropdown  list. Select the option “Coronavirus (8)”. The number “8” means that this group contains 8 genomes, which are from 8 herpesvirus strains. </p>

                <p><strong>Select a Genome. </strong>From the drop down list, select “Human SARS coronavirus 2 (SARS-CoV-2) isolate Wuhan-Hu-1”.</p>

                <p><strong>Ignore “Sequence  ID(s)”</strong> This is for querying one or a list of sequence IDs from the specified genome.  The IDs can be NCBI Protein Accession number, NCBI Protein GI, NCBI Gene ID, and  NCBI Locus Tag. </p>

                <p><strong>Ignore “Keywords” line.</strong> This is for querying one or more proteins based on a specific database ID (e.g., NCBI Locus Tag), gene symbol, or protein description.</p>

            <p><strong>Select Filter Options</strong> This allows you to set criteria to filter out proteins: </p>
            <ul>
                <li><strong>Select Subcellular Localization.</strong> This feature does not particularly fit in this virus use case. It is useful for bacterial vaccine target prediction. For bacterial vaccine development, it is often preferred to identify outer membrane or secreted proteins as vaccine targets.</li>
                <br>
                <li><strong>Maximum Number of Transmembrane Helices.</strong> It has been observed that a protein with more than one transmemebrane helix is hard to isolate from a recombinant <em>E. coli</em> strain. Therefore, this feature is useful if you consider using recombinant <em>E. coli</em> strains for protein isolation and purification.</li>
                <br>
                <li><strong>Minimum Adhesin Probability (0-1.0).</strong> An adhesin is a protein critical for helping a pathogen to enter a host cell. Neutralizing an adhesin is helpful for preventing a pathogen’s invasion. The default cutoff is 0.51.</li>
                <br>
                <li><strong>Have Orthologs in.</strong> This allows you to find conserved proteins among a selected list of strains. <strong>&nbsp;Tips:</strong> To select (or unselect) individual strains, press Ctrl (Cmd on Macs) and then select; For strain selection in continuous order, click a strain marking the beginning of your selection, then hold Shift while selecting the end of your selection.</li>
                <br>
                <li><strong>Exclude Proteins having Orthologs in Any of Selected  Genome(s).</strong> This is for excluding proteins that also also exist in a non-pathogenic strain(s). </li>
                <br>
                <li><strong>Similarity to Human Proteins.</strong> This determines whether vaccine targets that also exist in humans should be included, excluded, or not considered.</li>
                <br>
                <li><strong>Similarity to Mouse Proteins.</strong> This determines whether vaccine targets that also exist in mouse should be included, excluded, or not considered.</li>
                <br>
                <li><strong>Similarity to Pig Proteins.</strong> This determines whether vaccine targets that also exist in pigs should be included, excluded, or not considered.</li>
            </ul>
            
            <blockquote>
            <p><strong>NOTE: MHC  Class I &amp; II Epitope Prediction by Vaxitop.</strong> This option has been removed from current version of Vaxign2  query cover page. However, you can use the separate Vaxitop program to run it  name  replaces previously defined “Vaxitope” in order to avoid a naming conflict. </p>
            </blockquote>
            <p>For the purposes of the tutorial, do not specify any filter options.</p>
            <br>
            <p>
                <strong>Submit your query.</strong> Click the “<strong>Submit</strong>” button in the bottom of the Vaxign2 Query web interface.</p>
            <p>
            <br>

            <p><strong>Query result examination.</strong> After Vaxign2 finishes running the job, you will be redirected to the result page for your query. Thet total number of proteins identified will be displayed above the results table (Figure 3) which displays 50 proteins per page by default. You may change how many entries are shown per page as well as navigate between result pages below the results table. Each row of the table represents a specific protein which includes: (1) Protein Accession number, (2) Protein Name, (3) Gene Accession Number, (4) Gene Symbol, (5) Locus Tag, (6) Vaxign-ML score, (7) Localization, (8) Adhesin Probability, (9) Trans-membrane helices, (10) Similarity to Human Proteins, (11) Similarity to Mouse Proteins, and (12) Similarity to Pig Proteins.</p>

            <br>
            <div align="center">
                <figure>
                    <a href="{% static "tutorial/figure3.png" %}"><img src="{% static "tutorial/figure3.png" %}" alt="Figure 3: Vaxign2 results" width="800" /></a>
                    <figcaption>Figure 3: Vaxign2 results</figcaption>
                </figure>
            </div>
            
            <p><strong>Conservations among human coronaviruses</strong> Click “<strong>Analysis</strong>“ above the results table and then click “<strong>Show Genome Group [Coronavirus] Orthologs</strong>“. This will take you to a page that will allow you to select other coronaviruses to include in an ortholog table. Select all human coronaviruses, and then click “<strong>Submit</strong>“. An ortholog table will appear (Figure 4) showing that 10 proteins are conserved between the coronavirus genome group.</p>

            <div align="center">
                <figure>
                    <a href="{% static "tutorial/figure4.png" %}"><img src="{% static "tutorial/figure4.png" %}" alt="Figure 4: Vaxign2 ortholog results" width="800" /></a>
                    <figcaption>Figure 4: Vaxign2 ortholog results</figcaption>
                </figure>
            </div>

            <p><strong>Restriction of maximum number of Transmembrane Helices through filtering</strong> Back on the query result page, click “<strong>Filter Results</strong>“ above the results table. Once the filter frame appears, choose the “less than or equal to" (&lt;=) option in the dropdown field and enter “1“ in the text field to the right of the “Number of Transmembrane Helices“ label. This specifies a view of results for proteins with one or zero transmemebrane helices. To apply the filter, click the checkbox to the right of the quantity field and click “<strong>Submit</strong>“ at the top of the filter frame. The results will be updated, showing 19 proteins after filtering out 5 proteins that had more than one transmemebrane helix. Other filter rules may be applied in the filter frame.</p>

            <br>

            <p><strong>Sort results by Adhesin Probability.</strong> By default, Vaxign2 sorts results in descending order by Vaxign-ML Score. To change how the results are sorted, click the “<strong>Adhesin Probability</strong>“ column header once to sort the results by Adhesin Probability in ascending order, then once more for descending order.</p>

            <p><strong>Checking detailed results for one protein.</strong> The SARS-CoV-2 S protein (NCBI protein ID QHD43416|S) is a commonly used vaccine antigen in current COVID-19 vaccine development. In this example of Vaxign2 predicting a high adhesin probability and Vaxign-ML predicting the same protein as a good vaccine antigen with a score of 97.6, we will investigate the single protein in detail. Click the protein “<strong>QHD43416|S</strong>“. This will take you to a page with detailed predictions for the protein (Figure 5).</p>

            <p>From here, you may use the subnavigation to the left of the analysis window to switch between analyses tools.</p>

            <div align="center">
                <figure>
                    <a href="{% static "tutorial/figure5.png" %}"><img src="{% static "tutorial/figure5.png" %}" alt="Figure 5: Vaxign2 detailed protein results" width="800" /></a>
                    <figcaption>Figure 5: Vaxign2 detailed protein results</figcaption>
                </figure>
            </div>

            <p><strong>Vaxitop epitope prediction.</strong> Alleles may be selected with options for host, MHC class, MHC allele, and epitope length by selecting desired options in their respective dropdown menus and clicking “<strong>Add Allele</strong>“. Frequently used alleles may be added with a single click in the “Frequently used alleles“ section.</p>
            <p>Once the desired alleles have been selected, click “<strong>Show Vaxitop Results</strong>“ to populate the results table. Results may be filtered by p-value cutoff and a general search term for all fields in their respective text fields above the results table. Results may be copied to your clipboard or exported as a CSV, Excel file, or PDF file using the respective buttons directly above the results table.</strong></p>
            <p>By default, the table will only show 25 entries per results page. You may change this by selecting an alternative number in the dropdown menu below the bottom-left corner of the results table. You may view the next number of results according to your desired display interval by clicking on the the page number links or “<strong>Previous</strong>“ and “<strong>Next</strong>“ links to view the results page behind or ahead the page currently being observed.</p>

            <div align="center">
              <figure>
                  <a href="{% static "tutorial/figure6.png" %}"><img src="{% static "tutorial/figure6.png" %}" alt="Figure 6: Vaxitop prediction results" width="800" /></a>
                  <figcaption>Figure 6: Vaxitop prediction results</figcaption>
              </figure>
            </div>

            <p><strong>IEDB Epitopes.</strong> IEDB experimentally validated epitopes for B Cells or T Cells may be reviewed for a protein of interest. A results table will appear with information including IEDB ID, epitope, starting position, and ending location. Search, export, and results table navigation options behave the same as the Vaxitop prediction screen.</p>

            <div align="center">
              <figure>
                  <a href="{% static "tutorial/figure7.png" %}"><img src="{% static "tutorial/figure7.png" %}" alt="Figure 7: IEDB epitope results" width="800" /></a>
                  <figcaption>Figure 7: IEDB epitope results</figcaption>
              </figure>
            </div>

            <p><strong>IEDB Population Coverage.</strong> Global population coverage for MHC-I, MHC-II, as well as MHC-I and MHC-II combined may be selected in the dropdown menu in the upper left corner of the analysis screen. Map navigational, selection, and image download tools will appear at the top right corner of the map after the cursor is hovered over the map. Countries may be hovered over with the cursor to see coverage corresponding to the selected MHC option.</p>

            <div align="center">
              <figure>
                  <a href="{% static "tutorial/figure8.png" %}"><img src="{% static "tutorial/figure8.png" %}" alt="Figure 8: IEDB population coverage" width="800" /></a>
                  <figcaption>Figure 8: IEDB population coverage</figcaption>
              </figure>
            </div>

            <p><strong>EggNOG Functions.</strong> Functional annotations including Clusters of Orthologous Groups (COG) and Gene Ontology (GO) terms may be viewed on this analysis screen.</p>

            <div align="center">
              <figure>
                  <a href="{% static "tutorial/figure9.png" %}"><img src="{% static "tutorial/figure9.png" %}" alt="Figure 9: EggNOG functions" width="400" /></a>
                  <figcaption>Figure 9: EggNOG functions</figcaption>
              </figure>
            </div>

            <p><strong>EggNOG Orthologs.</strong> This screen displays orthologous proteins predicted by the EggNOG database in a Table View that may be viewed, navigated, and matched against a search term in a similar fashion to previous analysis screens. You may switch from Table View to a taxonomic Tree View by selecting the desired option in the dropdown box to the right of "Orthologous Protein Prediction"</p>

            <div align="center">
              <figure>
                  <a href="{% static "tutorial/figure10.png" %}"><img src="{% static "tutorial/figure10.png" %}" alt="Figure 10: EggNOG orthologs table view" width="400" /></a>
                  <figcaption>Figure 10: EggNOG orthologs table view</figcaption>
              </figure>
            </div>

            <div align="center">
              <figure>
                  <a href="{% static "tutorial/figure11.png" %}"><img src="{% static "tutorial/figure11.png" %}" alt="Figure 11: EggNOG orthologs tree view" width="600" /></a>
                  <figcaption>Figure 11: EggNOG orthologs tree view</figcaption>
              </figure>
            </div>

            <p><strong>Genome Group Ortholog Phylogeny.</strong> This screen displays a phylogenic tree for related organisms and orthologous proteins. A download feature appears in the top-right corner of the analysis screen after hovering the cursor over the tree. This feature is only available to logged in users.</p>

            <div align="center">
              <figure>
                  <a href="{% static "tutorial/figure12.png" %}"><img src="{% static "tutorial/figure12.png" %}" alt="Figure 12: Genome group ortholog phylogeny" width="800" /></a>
                  <figcaption>Figure 12: Genome group ortholog phylogeny</figcaption>
              </figure>
            </div>

            <p><strong><u>SECOND USE  CASE: <em>E. COLI</em> O157:H7 VACCINE TARGET  PREDICTION</u></strong></p>
            <p>We will not provide much detail for this use case here.<strong> </strong>The procedure to run this is similar  to what we explain above. One feature that is not demonstrated in the above  example is the Filter Option: </p>
            <ul>
            <li><strong>1. Select Subcellular Localization</strong>.  This option  allows you to select one or more subcellular locations. The options include: </li>
            <ul>
                <li>Any Localization </li>
                <li>Cell wall </li>
                <li>Cytoplasmic</li>
                <li>Cytoplasmic Membrane</li>
                <li>Excellular proteins</li>
                <li>Outer Membrane</li>
                <li>Periplasmic</li>
                <li>Unknown</li>
            </ul>
            </ul>
            <p>This filter  method is based on PSortb (<a href="http://www.psort.org/psortb/">http://www.psort.org/psortb/</a>),  which is specifically designed for prediction of subcellular localization of  proteins from Gram + or Gram – bacteria. It is not designed for other types of  microbes. </p>
            <p>Figure 7 provides some  concrete settings for an example analysis. In this example, we chose to  identify only those extracellular and outer membrane proteins. This analysis  results in 45 hits.  <strong> </strong><br />
            </p>
            <div>
                <div align="center">
                    <figure>
                        <a href="{% static "tutorial/figure13.png" %}"><img src="{% static "tutorial/figure13.png" %}" alt="Figure 13: Prediction of E. coli O157:H7 vaccine targets" width="800" /></a>
                        <figcaption>Figure 13: Prediction of E. coli O157:H7 vaccine targets</figcaption>
                    </figure>
                </div>
            <p>&nbsp;</p>
            </div>
            <p><strong><u>Register and Use a Vaxign2 Account for Predicted Result Storage and Sharing </u></strong><br />
            <strong>Purpose of having a Vaxign2  account.</strong> It is not required to open a Vaxign2/VIOLIN account. However, it provides some extra benefits for you: (1) It saves your dynamic analysis  results in the Vaxign2 system for your future reference and refinement. (2) It  promotes collaboration. You can share your Vaxign2 projects with your  colleagues, so the whole team can work on one Vaxign2 analysis project together. <strong></strong><br />
            <strong>Set up a Vaxign2 account. </strong>Click “<strong>My Analysis</strong>” on the left side  navigation bar in the Vaxign2 system (Figure 8). If you have not register,  please click on “<strong>Register an account</strong>”  and register using a web form. <br />
            <strong>Create a new project under  “My Analysis”.</strong> You can create a new project and follow the online  instruction. </p>
            <div align="center">
                <figure>
                    <a href="{% static "tutorial/figure14.png" %}"><img src="{% static "tutorial/figure14.png" %}" alt="Figure 14: Vaxign2 side navigation bar" width="150" /></a>
                    <figcaption>Figure 14: Vaxign2 side navigation bar</figcaption>
                </figure>
            </div>
            <br clear="all" />
            <p><strong>Exit Vaxign2.</strong> Once you complete your  exercises, you may log out of Vaxign2 if you generate your own account and log  in. To log out, navigate to the top of the window, and click on <strong>LOGOUT</strong>.</p>
            <p>&nbsp;</p>
            <p>------------------------</p>
            <p><strong>Note: </strong>Please see more information in our <a href="http://www.violinet.org/vaxign2/docs"><strong>Vaxign2 Documentation</strong></a> page. </p>
            <p>&nbsp;</p>
            <p><strong>Provide  Comments.</strong> Once you complete your exercises, you may log out of Vaxign2  if you generate your own account and log in. In an effort to improve the Vaxign2  performance and provide better support for the vaccine development community, we  would appreciate any comments/suggestions you may have regarding the Vaxign2  program.  To provide your comments, please email us or fill out a feedback online form in the Vaxign2 <a href="http://www.violinet.org/contact_us.php"><strong>Contact Us</strong></a> page. Thank you!! </p>
            <p>&nbsp;</p>(URL: <a href="http://www.violinet.org/vaxign2/vaxitop">http://www.violinet.org/vaxign2/vaxitop</a>).  Also, although the Vaxitop epitope prediction selection is not available in the  cover page, the Vaxitop epitope prediction method is available for any protein  you select after the initial screening. To predict MHC class I and II epitopes,  the Vaxitop method is based on prediction of position specific scoring matrices  (PSSM). Different from existing epitope prediction algorithms, Vaxitop relies  on statistical <em>P-</em>value (instead of a percentage or  top number) as the cutoff. A <em>P-</em>value of 0.05  provides a cutoff with high and balanced sensitivity and specificity. Under  this section, you can choose a<em> P-</em>value cutoff,  host species, MHC allele, and epitope length. It’s noted that the Vaxitop
            
            </div>
        </td>
    </tr>
</table>

{% include 'template_vaxign_bottom.html' %}

</body>
